File: examples/single_arm.py
"""End-to-end Python example: 10-second 1 kHz MIT control loop on a single
Damiao arm. Uses --mock for CI; --interface vcan0 for real hardware.

Usage:
    python examples/single_arm.py --mock
    python examples/single_arm.py --interface vcan0
"""

import argparse
import time

import numpy as np

import dm_control
from dm_control.damiao import DamiaoCodec, MotorType


def build_robot(use_mock: bool, interface: str) -> dm_control.Robot:
    transport = (
        dm_control.MockCanBus("vcan_mock")
        if use_mock
        else dm_control.SocketCanBus(interface, fd=False)
    )
    return (
        dm_control.RobotBuilder()
        .add_bus("main", transport, DamiaoCodec())
        .add_arm(
            "arm",
            bus="main",
            motors=[
                dm_control.MotorSpec("j0", MotorType.DM4340, send_id=0x01, recv_id=0x11),
                dm_control.MotorSpec("j1", MotorType.DM4340, send_id=0x02, recv_id=0x12),
                dm_control.MotorSpec("j2", MotorType.DM4340, send_id=0x03, recv_id=0x13),
                dm_control.MotorSpec("j3", MotorType.DM4340, send_id=0x04, recv_id=0x14),
            ],
        )
        .build()
    )


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("--mock", action="store_true", help="use MockCanBus (no hardware)")
    parser.add_argument("--interface", default="vcan0", help="SocketCAN interface")
    parser.add_argument("--seconds", type=float, default=10.0)
    parser.add_argument("--period", type=float, default=1e-3)
    parser.add_argument("--deadline-us", type=int, default=500)
    args = parser.parse_args()

    robot = build_robot(args.mock, args.interface)

    with robot:
        arm = robot["arm"]
        n = len(arm)
        cmds = np.zeros((n, 5), dtype=np.float64)
        cmds[:, 0] = 50.0  # kp
        cmds[:, 1] = 1.0  # kd

        deadline = time.monotonic() + args.seconds
        ticks = 0
        while time.monotonic() < deadline:
            t0 = time.monotonic()
            robot.tick(args.deadline_us)
            arm.mit_control(cmds)
            ticks += 1
            rem = args.period - (time.monotonic() - t0)
            if rem > 0:
                time.sleep(rem)
        print(f"completed {ticks} ticks")


if __name__ == "__main__":
    main()

File: crates/can-motor-control/examples/single_arm.rs
//! End-to-end Rust example: open a single Damiao arm on SocketCAN, run a
//! 10-second 1 kHz MIT control loop holding zero position, disable.
//!
//! Usage:
//!
//! ```bash
//! # one-time vcan0 setup (or use a real CAN interface):
//! sudo ip link add dev vcan0 type vcan
//! sudo ip link set vcan0 up
//!
//! cargo run --example single_arm -- vcan0
//!
//! # CAN-FD bus (interface must be FD-capable, e.g. `ip link set <iface> mtu 72`):
//! cargo run --example single_arm -- canfd0 --fd
//! ```

use std::env;
use std::time::{Duration, Instant};

use can_motor_control::{
    CanBus, MitCmd, MotorCodec, MotorSpec, MotorTypeId, RobotBuilder, SocketCanBus,
};
use damiao_codec::{parse_motor_type, DamiaoCodec, DamiaoMotorType};

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let _ = parse_motor_type; // silence import for the curious reader.

    // Args: an optional positional interface name and an optional `--fd` flag.
    let mut interface = "vcan0".to_string();
    let mut fd = false;
    for arg in env::args().skip(1) {
        if arg == "--fd" {
            fd = true;
        } else {
            interface = arg;
        }
    }
    println!("opening SocketCAN interface: {interface} (fd={fd})");

    let transport: Box<dyn CanBus> = Box::new(SocketCanBus::open(&interface, fd)?);
    let codec: Box<dyn MotorCodec> = Box::new(DamiaoCodec::new());

    let mut robot = RobotBuilder::new()
        .add_bus("main", transport, codec)
        .add_arm(
            "arm",
            "main",
            vec![
                MotorSpec::new("j0", MotorTypeId::from(DamiaoMotorType::DM4340), 0x01, 0x11),
                MotorSpec::new("j1", MotorTypeId::from(DamiaoMotorType::DM4340), 0x02, 0x12),
                MotorSpec::new("j2", MotorTypeId::from(DamiaoMotorType::DM4340), 0x03, 0x13),
                MotorSpec::new("j3", MotorTypeId::from(DamiaoMotorType::DM4340), 0x04, 0x14),
            ],
        )
        .build()?;

    println!("connecting...");
    robot.connect()?;
    println!("enabling...");
    robot.enable()?;

    let deadline_per_tick = Duration::from_micros(500);
    let loop_period = Duration::from_millis(1);
    let start = Instant::now();

    println!("running 10s control loop (1 kHz) holding zero position via MIT...");
    while start.elapsed() < Duration::from_secs(10) {
        let tick_start = Instant::now();
        robot.tick(deadline_per_tick)?;
        let arm = robot
            .group_mut("arm")
            .and_then(|g| g.as_arm_mut())
            .expect("arm group");
        let n = arm.len();
        let cmds: Vec<MitCmd> = (0..n)
            .map(|_| MitCmd {
                kp: 50.0,
                kd: 1.0,
                q: 0.0,
                dq: 0.0,
                tau: 0.0,
            })
            .collect();
        arm.mit_control(&cmds)?;
        if let Some(rem) = loop_period.checked_sub(tick_start.elapsed()) {
            std::thread::sleep(rem);
        }
    }

    println!("disabling...");
    robot.disable()?;
    println!("done.");
    Ok(())
}
